File: pkg/prefix_sum/path_sum.go
package prefixsum

import (
	. "gocode/types"
)

/**
 * 437. Path Sum III
 *
 * Given the root of a binary tree and an integer targetSum,
 * return the number of paths where the sum of the values along the path equals targetSum.
 *
 * "traveling only from parent nodes to child nodes)""
 *
 * Time: O(n), Space: O(n) — prefix map
 */
func pathSum(root *TreeNode, targetSum int) int {
	// prefixCount[sum] = number of paths from root with this prefix sum
	prefixCount := map[int]int{0: 1} // base case: empty path
	res := 0

	var dfs func(node *TreeNode, currSum int)
	dfs = func(node *TreeNode, currSum int) {
		if node == nil {
			return
		}

		currSum += node.Val

		// how many paths ending at current node sum to targetSum?
		// currSum - targetSum = prefix sum we need to have seen before
		res += prefixCount[currSum-targetSum]

		// record current prefix sum
		prefixCount[currSum]++

		dfs(node.Left, currSum)
		dfs(node.Right, currSum)

		// undo — remove current node's prefix sum when backtracking
		prefixCount[currSum]--
	}

	dfs(root, 0)
	return res
}

File: python/prefix_sum/path_sum.py
from collections import defaultdict
import sys, os
from typing import Optional
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))

from common import TreeNode

class Solution:
    """
    437. Path Sum III

    prefixSum[j] - prefixSum[i] = k => prefixSum[i] = prefixSum[j] - k
    """
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        prefix_counts = defaultdict(int)
        prefix_counts[0] = 1   # empty path base case

        def dfs(node, curr_sum) -> int:
            if not node:
                return 0

            curr_sum += node.val

            # how many paths ending here sum to target
            count = prefix_counts[curr_sum - targetSum]

            # explore children with updated prefix
            prefix_counts[curr_sum] += 1
            count += dfs(node.left, curr_sum)
            count += dfs(node.right, curr_sum)
            prefix_counts[curr_sum] -= 1   # backtrack

            return count

        return dfs(root, 0)